pkg/watch/health.go

```diff
@@ -0,0 +1,240 @@
+// longer be running.
+func WatchHealth(node, consul, authtoken string, logger *logging.Logger, shutdownCh chan struct{}) {
+	var store kp.Store
+	tochan := make(chan bool)
+	pods := []*PodWatch{}
+	if authtoken != "" {
+		store = kp.NewConsulStore(kp.Options{
+			Address: consul,
+			HTTPS:   false,
+			Token:   authtoken,
+			Client:  net.NewHeaderClient(nil, http.DefaultTransport),
+		})
+	} else {
+		store = kp.NewConsulStore(kp.Options{
+			Address: consul,
+			HTTPS:   false,
+			Client:  net.NewHeaderClient(nil, http.DefaultTransport),
+		})
+	}
+
+	go startTimer(tochan, POLL_KV_FOR_PODS)
+	for {
+		select {
+		case _ = <-tochan:
+			// start timer again
+			go startTimer(tochan, POLL_KV_FOR_PODS)
+
+			// check if pods have been added or removed
+			// starts monitor routine for new pods
+			// kills monitor routine for removed pods
+			pods = updateHealthMonitors(store, pods, node, logger)
+		case _ = <-shutdownCh:
+			return
+		}
+	}
+}
+
+// Monitor Health is a go routine that runs as long as the
+// service it is monitoring. Every HEALTHCHECK_INTERVAL it
+// performs a health check and writes that information to
+// consul
+func (p *PodWatch) MonitorHealth(node string, store kp.Store, shutdownCh chan bool) {
+	statusCheck := fmt.Sprintf(kp.GetStatusCheck(), p.manifest.Manifest.StatusPort)
+	tochan := make(chan bool)
+
+	go startTimer(tochan, HEALTHCHECK_INTERVAL)
+	for {
+		select {
+		case _ = <-tochan:
+			p.checkHealth(statusCheck, node, store)
+			go startTimer(tochan, HEALTHCHECK_INTERVAL)
+		case _ = <-shutdownCh:
+			return
+		}
+	}
+}
+
+func (p *PodWatch) checkHealth(healthCheck, node string, store kp.Store) {
+	healthstate, res, err := check(healthCheck)
+	health := health.Result{
+		ID:      p.manifest.Manifest.Id,
+		Node:    node,
+		Service: p.manifest.Manifest.Id,
+		Status:  healthstate,
+		Output:  res,
+	}
+
+	err = writeToConsul(health, store)
+	if err != nil {
+		p.logger.WithField("CONSUL WRITE ERROR", err).Fatalln("failed to write health data to consul")
+	}
+}
+
+// check is invoked periodically and runs the health check
+// string c as a shell script
+func check(c string) (health.HealthState, string, error) {
+	output := new(bytes.Buffer)
+	cmd, err := RunScript(c)
+
+	cmd.Stdout = output
+	cmd.Stderr = output
+	err = cmd.Start()
+	_ = cmd.Wait()
+	if cmd.ProcessState.Success() == true {
+		return health.Passing, output.String(), err
+	} else {
+		return health.Critical, output.String(), err
+	}
+}
+
+// once we get health data we need to make a put request
+// to consul to put the data in the KV Store
+func writeToConsul(res health.Result, store kp.Store) error {
+	var value string
+	// key =  service/node
+	// if status == passing: value = status
+	// else: values = health.Status/health.Output
+	key := fmt.Sprintf("%s/%s", res.Service, res.Node)
+	if res.Status == health.Passing {
+		value = "passing"
+	} else {
+		value = fmt.Sprintf("%s/%s", res.Status, res.Output)
+	}
+	_, err := store.Put(key, value)
```

Comment:
Maybe an alternative would be to move the health structs in `health` to `kp` and make them part of the interface to `Store`. Specifically something like:

```
func writeToConsul(res kp.healthResult, store kp.Store) error {
   store.SetHealth(res)
}
```
This lets the underlying schema in Consul be managed by the `kp` package directly